### PR TITLE
Clearing attribute buffers after an elements fails to parse

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**

### CODE REVIEW

File: Test/FMI3/fmi3_import_terminals_and_icons_test.cpp [hunk 1]
```diff
@@ -116,3 +116,35 @@ TEST_CASE("Error check; Terminals with duplicate names; edge case of empty name"
...
+    REQUIRE(fmi3_import_get_has_terminals_and_icons(fmu) == 1);
+
+    REQUIRE(fmi3_testutil_log_contains(tfmu, "Parsing XML element 'Terminal': required attribute 'name' not found"));
+
+    fmi3_import_terminal_t* term = fmi3_import_get_terminal_by_name(fmu, "terminalA");
+    REQUIRE(term != nullptr);
+    REQUIRE_STREQ(fmi3_import_get_terminal_name(term), "terminalA");
+    // TODO: Check that optional attributes from Terminal without name are not present in this one
+
+    // TODO: Current example xml includes elements/attributes not yet parsed
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_terminals_and_icons_test.cpp:147]: I guess not part of this ticket?

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 2]
```diff
@@ -1320,6 +1320,13 @@ static void XMLCALL fmi3_parse_element_end(void* c, const char *elm) {
 
     if (context->skipElementCnt) {
         context->skipElementCnt--;
+        // Need to clear buffer
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_parser.c:1323]: Wouldn't it make more sense to move this code to the parse_element_start code? In the block where we detect that the handler failed.

**Review by filip-stenstrom — COMMENTED** [review]
Looks good in general, just see comments. Good with the extra fixes as well.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 3]
```diff
@@ -1320,6 +1320,13 @@ static void XMLCALL fmi3_parse_element_end(void* c, const char *elm) {
...
+        // else: non-parsed attributes of elements that failed to parse will be parsed with next element
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c]: Taking the suggestion on the comment. 

As for renaming the functions, this is a bit odd, since there is already a `fmi3_xml_get_attr_str` function, and the difference to `fmi3_xml_parse_attr_as_string` are the return types, string(const char**) vs jm_vector(char)*.

So... the most accurate rename would be `fmi3_xml_get_attr_as_jm_vector_char`, which doesn't quite feel right. 

One could look into the question if one of these 2 functions is actually redundant, but I would rather take such cleanup as a separate issue.

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 2 above)
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c]: Good point, it does make more sense. Moved it.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 4]
```diff
@@ -1320,6 +1320,13 @@ static void XMLCALL fmi3_parse_element_end(void* c, const char *elm) {
...
+        for (int i = 0; i < fmi3_xml_attr_number; i++) {
+            if (jm_vector_get_item(jm_string)(context->attrMapById, i)) {
+                jm_vector_set_item(jm_string)(context->attrMapById, i, 0);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c]: This was not duplicated for terminalsAndIcons, hence the change only exists once.

File: Test/FMI3/fmi3_import_error_handling_model_structure.cpp [hunk 5]
```diff
@@ -234,3 +234,17 @@ TEST_CASE("EventIndicators; multiple attribute issues") {
     REQUIRE(fmi3_testutil_get_num_problems(tfmu) == 2);
     fmi3_testutil_import_free(tfmu);
 }
+
+TEST_CASE("Buffer clearing; attributes from invalid elements need to be cleared") {
+    const char* xmldir = FMI3_TEST_XML_DIR "/error_handling/model_structure/buffer_clearing";
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply Test/FMI3/fmi3_import_error_handling_model_structure.cpp]: Good idea, done.

File: Test/FMI3/fmi3_import_error_handling_model_variables.cpp [hunk 6]
```diff
@@ -143,12 +140,12 @@ TEST_CASE("Test missing required attribute plus other attribute errors/warnings"
...
     REQUIRE(fmi3_testutil_log_contains(tfmu, "Parsing XML element 'Float64': required attribute 'name' not found"));
 
-    // We do not check attribute errors if name/valueReference are erroneous
+    // We do not check attribute errors if name/valueReference are erroneous; test absence of error message
     REQUIRE(!fmi3_testutil_log_contains(tfmu, "A variable must not refer to itself in the attribute 'previous'."));
 
     REQUIRE(fmi3_testutil_log_contains(tfmu, "Fatal failure in parsing ModelVariables. Variable(s) failed to parse or an essential error check failed.")); // counts as 2
 
-    REQUIRE(fmi3_testutil_get_num_problems(tfmu) == 4);
+    REQUIRE(fmi3_testutil_get_num_problems(tfmu) == 3);
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply Test/FMI3/fmi3_import_error_handling_model_variables.cpp:148]: The return value changes in fmi3_xml_variable.c.

Zero (rather than negative) returns would trigger the verbose buffer cleanup of non-parsed attributes. I.e., `Attribute '%s' not processed by element '%s' handle"`, where I believe the intention is to identify attributes that may be misplaced for a variable.

File: Test/FMI3/fmi3_import_terminals_and_icons_test.cpp (hunk 1 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/fmi3_import_terminals_and_icons_test.cpp:147]: Correct, we don't process any attributes here yet to be able to test this.

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 7]
```diff
@@ -2201,7 +2201,7 @@ int fmi3_xml_handle_Enumeration(fmi3_xml_parser_context_t *context, const char*
     /* Extract common Variable info & handle errors*/
     if (fmi3_xml_handle_Variable_error_check_wrapper(context, data)) {
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_variable.c]: Good suggestion, I adjusted accordingly and like the result, see changes.

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_parser.c]: Suggestion: Okay good.

Renaming: Yes, let's take that another time. It would require some more thinking through. I mainly wanted to express that it's something we probably should do in the future.

File: Test/FMI3/fmi3_import_error_handling_model_structure.cpp [hunk 8]
```diff
@@ -234,3 +234,17 @@ TEST_CASE("EventIndicators; multiple attribute issues") {
     REQUIRE(fmi3_testutil_get_num_problems(tfmu) == 2);
     fmi3_testutil_import_free(tfmu);
 }
+
+TEST_CASE("Attribute fuffer clearing; attributes from invalid elements need to be cleared") {
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_error_handling_model_structure.cpp:238]:
```suggestion
TEST_CASE("Attribute buffer clearing; attributes from invalid elements need to be cleared") {
```

**Review by filip-stenstrom — APPROVED** [review]
